share/new_eds/proofs_cache.go

```diff
@@ -0,0 +1,276 @@
+package eds
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"sync/atomic"
+
+	"github.com/ipfs/boxo/blockservice"
+	blocks "github.com/ipfs/go-block-format"
+	"github.com/ipfs/go-cid"
+
+	"github.com/celestiaorg/celestia-app/pkg/wrapper"
+	"github.com/celestiaorg/nmt"
+	"github.com/celestiaorg/rsmt2d"
+
+	"github.com/celestiaorg/celestia-node/share"
+	"github.com/celestiaorg/celestia-node/share/ipld"
+	"github.com/celestiaorg/celestia-node/share/shwap"
+)
+
+var _ Accessor = (*proofsCache)(nil)
+
+// proofsCache is eds accessor that caches proofs for rows and columns. It also caches extended
+// axis Shares. It is used to speed up the process of building proofs for rows and columns,
+// reducing the number of reads from the underlying accessor.
+type proofsCache struct {
+	inner Accessor
+
+	// lock protects axisCache
+	lock sync.RWMutex
+	// axisCache caches the axis Shares and proofs. Index in the slice corresponds to the axis type.
+	// The map key is the index of the axis.
+	axisCache []map[int]axisWithProofs
```

Comment:
Is there a reason to make it a map with an int key instead of a slice? We could speed up reads by using a slice there.

In the future, if that becomes a contention point, we could make a lock per axis instead of the whole cache.

Reply:
Few minor advantages for map:
- it allows nice (val, ok) check. Otherwise we would need to make cache an array of pointers to check for nil. `[][]*axisWithProofs`
- it grows gradually with cache size, so there is no need to allocate array for all axis right away.

Reply:
> Otherwise we would need to make cache an array of pointers to check for nil. 

That's fine, that would also allow gradual growth of the cache

Reply:
Doubt mutex it can become contention point, it locks on writing to map and there are at most 256 writes in lifespan of cached accessor.